SwanCustomEnvironments/swancustomenvironments/serverextension.py

```diff
@@ -0,0 +1,88 @@
+from tornado import web
+from traitlets.config import Configurable
+
+from jupyter_server.base.handlers import JupyterHandler, APIHandler
+from jupyter_server.utils import url_path_join
+
+import subprocess
+from os import environ, path
+
+
+class SwanCustomEnvironments(Configurable):
+    # TODO set configs here, check e.g. SwanShare
+    pass
+
+
+class SwanCustomEnvironmentsApiHandler(APIHandler):
+
+    config = None
+
+    def initialize(self):
+        self.config = SwanCustomEnvironments(config=self.config)
+
+    @web.authenticated
+    def get(self):
+        self.set_header("Content-Type", "text/event-stream")
+        
+        env_name = self.get_query_argument("env", default=None)
+        req = self.get_query_argument("req", default=None)
+        clear = self.get_query_argument("clear", default="true")
+        accpy_version = self.get_query_argument("accpy", default=None)
+        custom_python = self.get_query_argument("python", default=None)
+
+        if not req.startswith("http"):
+            req = path.join(environ["HOME"], req)
+
+        try:
+            arguments = ["--env", env_name, "--req", req]
+            if clear.lower() == "true":
+                arguments.extend(["--clear"])
+            if accpy_version is not None:
+                arguments.extend(["--accpy", accpy_version])
+            if custom_python is not None:
+                arguments.extend(["--python", custom_python])
+            
+            makenv_process = subprocess.Popen(["/srv/singleuser/bin/makenv", *arguments], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+            
+            for line in iter(makenv_process.stdout.readline, b""):
+                self.write(f"data: {line.decode('utf-8')}\n\n")
```

Comment:
Where does this write?

Reply:
This communication is done through a protocol named SSE that works on top of HTTP. That line is (somehow, cause idk the details of the method) sending that to the frontend, with the help of the flush method that needs to be called to keep the frontend updating the "writes" that it gets

Reply:
But does this mean that the `write` method of an `APIHandler` writes to a websocket? Which one? Where is it configured?

Reply:
Im no longer using websockets. Now, the client makes one request to the server and waits for multiple responses, until something specific and pre defined is sent and the connection closes (see customenvs.html)

Reply:
Ok, let me comment below then.